### ISSUE TITLE
get_parameters() vs. parameters[-1] 

### ISSUE THREAD

**Issue by ulzegasi** [eth-cscs/abcpy#36]
Hi there, I've got an issue with the function get_parameters(). I am inferring three parameters (a1, a2, eps2) that are initially uniformly distributed in given intervals.
I run the following:

```
import numpy as np
import statistics
import time
from IterMapModel import IterMap

# Load data
data = [np.loadtxt('synth_dat_test.dat',dtype=float)]

# Define Graphical Model
from abcpy.continuousmodels import Uniform
lp = len(data[0])
a1 = Uniform([[1.3], [1.7]], name='a1')
a2 = Uniform([[1.6], [2.1]], name='a2')
eps2 = Uniform([[0.2], [0.5]], name='eps2')
timeseries = IterMap([lp, a1,a2,eps2], name = 'timeseries')

# Define backend
from abcpy.backends import BackendDummy as Backend
backend = Backend()

# Define Statistics
from abcpy.statistics import Identity
statistics_calculator = Identity(degree=2, cross=False)

from abcpy.summaryselections import Semiautomatic
summary_selection = Semiautomatic([timeseries], statistics_calculator, backend, n_samples = 10, n_samples_per_param = 1, seed = 1)
statistics_calculator.statistics = lambda x, f2 = summary_selection.transformation, f1 = statistics_calculator.statistics: f2(f1(x))

# Define distance
from abcpy.distances import Euclidean
distance_calculator = Euclidean(statistics_calculator)

# Define kernel
from abcpy.perturbationkernel import DefaultKernel
kernel = DefaultKernel([a1,a2,eps2])

# SABC ##
tsabc=time.time()
from abcpy.inferences import SABC
sampler = SABC([timeseries], [distance_calculator], backend, kernel, seed = 1)
steps, epsilon, n_samples, n_samples_per_param = 5, 100, 20, 1
tstart = time.time()
journal_sabc = sampler.sample([timeseries_obs], steps, epsilon, n_samples, n_samples_per_param,full_output=0)

samples = journal_sabc.parameters[-1]
#### Posterior sample for parameter a1 obtained using parameters[-1] #### 
print("Parameters[-1] gives for parameter a1: ")
print(samples[:,0])
print("Posterior mean from parameters[-1] is ")
print(statistics.mean(samples[:,0]))
#### Posterior sample for parameter a1 obtained using get_parameters() #### 
a1_post_sample, a2_post_sample, eps2_post_sample, post_weights = \
    np.array(journal_sabc.get_parameters()['a1']), np.array(journal_sabc.get_parameters()['a2']), \
    np.array(journal_sabc.get_parameters()['eps2']), np.array(journal_sabc.get_weights())
print("Get_parameters() gives for parameter a1: ")
print(a1_post_sample.flatten())
print("Posterior mean from get_parameters() is ")
print(statistics.mean(a1_post_sample.flatten()))

print("Posterior_mean() gives for a1: ")
print(journal_sabc.posterior_mean()[0])
```

The output I get is as follows,

```
Parameters[-1] gives for parameter a1: 
[1.68659718 1.39658876 1.50724835 1.53917481 1.30862889 1.50724835
 1.39658876 1.68659718 1.62423885 1.48039411 1.6568104  1.39658876
 1.62423885 1.50724835 1.41777341 1.68130142 1.6568104  1.39658876
 1.39658876 1.63262029]
Posterior mean from parameters[-1] is 
1.5249937330687782
Get_parameters() gives for parameter a1: 
[1.41777341 1.68130142 1.49302175 1.50724835 1.48039411 1.53917481
 1.63262029 1.50282747 1.44300723 1.6568104  1.47250143 1.39658876
 1.49733069 1.57906456 1.30862889 1.66946639 1.62423885 1.65467304
 1.38761389 1.68659718]
Posterior mean from get_parameters() is 
1.5315441460199923
Posterior_mean() gives for a1: 
1.5249937330687782
```
The posterior parameter samples obtained with parameters[-1] DIFFER from the parameter samples obtained with the function get_parameters(). 
IMPORTANT: this difference exists only if 'full_output' = 0 AND 'steps' is odd (3,5,7,...). However, if I set 'full_output' = 1 OR 'steps' is even (2,4,6,...) then the two posterior parameter samples coincide. 

Is there any explanation for this "odd" behaviour? 

The mean value of the parameter samples obtained with parameter[-1] is the same as the mean provided by the function posterior_mean(). It seems that, while parameters[-1] behaves correctly, there might be some problems with get_parameters().

I would appreciate some help!
Thanks a lot!

**Comment by statrita2004** [comment]
Hi Simone, We have fixed the bug in our recent release 0.5.1. Can you please test it and let us know?

**Comment by ulzegasi** [comment]
Hi Rito, sure, I’m very busy today, I’ll be on it next week.
Have a nice weekend!
Simone

**Comment by statrita2004** [comment]
The bug has been fixed in release 0.5.2. So I am closing this issue. Thanks a lot for pointing the issue.
